backend/internal/service/handler/auth.go

```diff
@@ -0,0 +1,155 @@
+package handler
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"platnm/internal/errs"
+	"sync"
+	"time"
+
+	"net/http"
+
+	"github.com/gofiber/fiber/v2"
+	"github.com/google/uuid"
+	"golang.org/x/oauth2"
+)
+
+var (
+	SpotifyAuthURL  = "https://accounts.spotify.com/authorize"
+	SpotifyTokenURL = "https://accounts.spotify.com/api/token"
+)
+
+type TokenResponse struct {
+	AccessToken  string `json:"access_token"`
+	TokenType    string `json:"token_type"`
+	Scope        string `json:"scope"`
+	ExpiresIn    int    `json:"expires_in"`
+	RefreshToken string `json:"refresh_token"`
+}
+
+// this is used to keep track of associated code verifier for each state since code verifier will be needed again to get token
+type CodeVerifierStore struct {
+	CodeVerifierMap map[string]string
+	Mu              sync.Mutex
+}
+
+type AuthHandler struct {
+	CodeVerifierStore *CodeVerifierStore
+}
+
+func NewAuthHandler() *AuthHandler {
+	return &AuthHandler{
+		CodeVerifierStore: &CodeVerifierStore{
+			CodeVerifierMap: make(map[string]string),
+		},
+	}
+}
+
+func (h *AuthHandler) SpotifyRedirect(c *fiber.Ctx) error {
+	verifier := oauth2.GenerateVerifier()
+	challenge := oauth2.S256ChallengeFromVerifier(verifier)
+
+	// store state as uuid for now
+	state := uuid.NewString()
+
+	ctx, cancel := context.WithTimeout(context.TODO(), 1*time.Minute)
+
+	h.CodeVerifierStore.Mu.Lock()
+	h.CodeVerifierStore.CodeVerifierMap[state] = verifier
+	h.CodeVerifierStore.Mu.Unlock()
+
+	go func() {
+		<-ctx.Done()
+		h.CodeVerifierStore.Mu.Lock()
+		delete(h.CodeVerifierStore.CodeVerifierMap, state)
+		h.CodeVerifierStore.Mu.Unlock()
+		cancel()
+		fmt.Printf("state %s has been deleted\n", state)
+	}()
+
+	params := url.Values{}
+	params.Add("client_id", os.Getenv("SPOTIFY_CLIENT_ID"))
+	params.Add("response_type", "code")
+	params.Add("redirect_uri", os.Getenv("SPOTIFY_REDIRECT_URI"))
+	params.Add("state", state)
+	params.Add("scope", "user-read-private")
+	params.Add("code_challenge_method", "S256")
+	params.Add("code_challenge", challenge)
+
+	u, err := url.Parse(SpotifyAuthURL)
+	if err != nil {
+		return errs.InternalServerError()
```

Comment:
```suggestion
		return err
```
`errs.ErrorHandler` will fail to cast into an `errs.HTTPError` so it will become an `errs.InternalServerError()` BUT will hit the `slog.Error(...)` line with the actual err info